### PR TITLE
Add phi3 128K model support

### PR DESCRIPTION
**Pull request by liuwei-git**
ref #6849 

The only difference between phi3 4k and 128k model is from the rotary embedding. 128k model adds long/short rope scaling factors (freq_factors) and an attn factor to each hidden dimension. The chosen of long/short factor is based on the total length of the input sequences, i.e, the kv context size.

```
seq_len = torch.max(position_ids) + 1
if seq_len > self.original_max_position_embeddings:
    ext_factors = torch.tensor(self.long_factor, dtype=torch.float32, device=x.device)
else:
    ext_factors = torch.tensor(self.short_factor, dtype=torch.float32, device=x.device)

inv_freq_shape = torch.arange(0, self.dim, 2, dtype=torch.int64, device=x.device).float() / self.dim
self.inv_freq = 1.0 / (ext_factors * self.base**inv_freq_shape)
```

The attn factor value is based on the postional embedding size.
```
scale = self.max_position_embeddings / self.original_max_position_embeddings
if scale <= 1.0:
    scaling_factor = 1.0
else:
    scaling_factor = math.sqrt(1 + math.log(scale) / math.log(self.original_max_position_embeddings))
```

## Workflow
* convert-hf-to-gguf.py: Write long/short freq factors to gguf metadata for phi3 model
* llama.cpp: 
    - load the freq factors and attn factor from metadata
    - take freq factors as an input tensor of phi3 model, and a source of k/q rope tensor
    - choose the long or short freq_factors based on the context size when setting the tensor value
    
* ggml: update rope op to support long/short freq factors:
    - [x] CPU
    - [x] CUDA
    - [x] Metal
    - [ ] SYCL
    - [ ] Vulkan

## Test
* https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/tree/main
    - clone the model and use `convert-hf-to-gguf.py` to convert to gguf format.
    - use passkey to test the model, e.g,:
```
    passkey phi3_128k_fp16.gguf 500
```

### CODE REVIEW

**Review by ggerganov — COMMENTED** [review]
Nice! I'll add the Metal support in a day or two if it is not yet pushed

**Comment by liuwei-git** [comment]
Thanks @ggerganov for your help. I did not have device to test metal, so not implement that part.

**Comment by ggerganov** [comment]
Looking into this now

**Comment by liuwei-git** [comment]
> I would like to refactor the `ggml_rope_custom` API and remove `ggml_rope_with_freq_factors` before merging - will push in a bit

The  refactor make the api looks more clean, truly great.

**Comment by slaren** [comment]
I would prefer if the scaling factors were exported as a tensor rather than metadata, it would remove quite a bit of code and it would be more efficient.

**Comment by ggerganov** [comment]
Yup, would be better to have the factors as tensors. @liuwei-git would you like to give this a go?

File: llama.cpp [hunk 1]
```diff
@@ -11165,6 +11241,23 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
...
+    if (lctx.freq_factors) {
+        // TODO: this might have to be hparams.n_rot instead of hparams.n_embd_head_k, but maybe it does not matter
+        const auto freq_dim = hparams.n_embd_head_k / 2;
+
+        GGML_ASSERT(lctx.freq_factors->ne[0] == freq_dim);
+        GGML_ASSERT(hparams.rope_long_factors.size()  == freq_dim);
+        GGML_ASSERT(hparams.rope_short_factors.size() == freq_dim);
+
+        // choose long/short freq factors based on the context size
+        const auto n_ctx = llama_n_ctx(&lctx);
```
Comment (slaren) [inline llama.cpp:11253]: Would this work correctly with multiple sequences? Maybe something like `llama_n_ctx(&lctx) / llama_n_seq_max(&lctx)` would be correct in more cases, but still not in every case.

Comment (compilade) [reply llama.cpp:11253]: > Maybe something like `llama_n_ctx(&lctx) / llama_n_seq_max(&lctx)`

For Transformer-like models, this would always equal 1.

`llama_n_ctx(&lctx) / cparams.n_seq_max` would be what you meant.

**Review by ggerganov — APPROVED** [review]
Should be good to merge

**Comment by slaren** [comment]
I think this is going to cause the rope to be run on the CPU always, because the scheduler prefers running ops that use weights in the backend on the weights. I will fix that after this is merged.

**Comment by ggerganov** [comment]
Ok. Btw, do you see something that could affect the performance of `phi-2` (no rope factors)? The benchmark is half the performance than usual (217 req/s) and I'm wondering if it is a fluke, because I don't reproduce on my RTX 2060

**Comment by slaren** [comment]
Looking at the graphs, it seems that the load time increased, but the throughput looks similar. Maybe it was a fluke?

**Comment by dillfrescott** [comment]
the model seems to be doing rather poorly. I cannot tell if its a tokenizer issue or just  the model itself, but I quantized the 128k medium instruct model to a q8_0 and its failing pretty simple logic issues. Perhaps its just not good with rather basic math?

I tried a temperature of `1` down to `0.6` and even down to `0` and its still not fairing well on logical questions. I was expecting more from a phi model, which leads me to think it may be some other underlying issue.

**Comment by RonanKMcGovern** [comment]
Did you try running a 16 bit gguf model and seeing how that performs?

**Comment by AlessandroW** [comment]
> Did you try running a 16 bit gguf model and seeing how that performs?

I tried the prompt on the 128k small instruct model (f16 and Q4_K_M from https://huggingface.co/AlessandroW/Phi-3-mini-128k-instruct-gguf) and both models performed similar to an older ChatGPT version https://neil.fraser.name/news/2023/02/17/. The quantized model replied:
```
<|user|> There is a roll of tape. The tape is 100 meters long when unrolled. When rolled up, the outer diameter is 10 cm, and the inner diameter is 5 cm. How thick is the tape?<|end|><|assistant|> To find the thickness of the tape, we need to calculate the volume of the tape when it is rolled up and then use that to find the thickness.

First, let's calculate the volume of the cylinder formed by the outer diameter. The formula for the volume of a cylinder is V = πr²h, where r is the radius and h is the height (or length in this case).

The outer radius (r_outer) is half of the outer diameter, so r_outer = 10 cm / 2 = 5 cm.

The length of the tape (h) is 100 meters, but we need to convert it to centimeters because the radius is in centimeters. So, h = 100 meters * 100 cm/meter = 10000 cm.

Now, let's calculate the volume of the outer cylinder (V_outer):
V_outer = π * (r_outer)² * h
V_outer = π * (5 cm)² * 10000 cm
V_outer = π * 25 cm² * 10000 cm
V_outer = 250000π cm³

Next, let's calculate the volume of the inner cylinder (V_inner), which represents the empty space inside the tape. The inner radius (r_inner) is half of the inner diameter, so r_inner = 5 cm / 2 = 2.5 cm.

The volume of the inner cylinder (V_inner) is:
V_inner = π * (r_inner)² * h
V_inner = π * (2.5 cm)² * 10000 cm
V_inner = π * 6.25 cm² * 10000 cm
V_inner = 62500π cm³

The volume of the tape itself (V_tape) is the difference between the outer and are not standard mathematical operations, and thus are not applicable in this context.

Since the question seems to be asking for the volume of the tape itself, we will subtract the inner volume from the outer volume to find the volume of the tape:

V_tape = V_outer - V_inner
V_tape = 250000π cm³ - 62500π cm³
V_tape = 187500π cm³

To get the numerical value, we use the approximation π ≈ 3.14159:

V_tape ≈ 187500 * 3.14159 cm³
V_tape ≈ 588746.25 cm³

Therefore, the volume of the tape itself is approximately 588746.25 cubic centimeters.<|endoftext|>
```
The f16 model replied
```
<|user|> There is a roll of tape. The tape is 100 meters long when unrolled. When rolled up, the outer diameter is 10 cm, and the inner diameter is 5 cm. How thick is the tape?<|end|><|assistant|> To find the thickness of the tape, we need to calculate the volume of the tape when it is rolled up and then use that to find the thickness.

First, let's find the volume of the entire length of the tape if it were a cylinder with the outer diameter. The formula for the volume of a cylinder is V = πr²h, where r is the radius and h is the height (or length in this case).

The outer radius (R) is half of the outer diameter, so R = 10 cm / 2 = 5 cm.

The length (h) of the tape is 100 meters, which we need to convert to centimeters because the diameter is in centimeters. 100 meters = 100 * 100 cm = 10000 cm.

Now, let's calculate the volume of the outer cylinder (V_outer):
V_outer = π * R² * h
V_outer = π * (5 cm)² * 10000 cm
V_outer = π * 25 cm² * 10000 cm
V_outer = 250000π cm³

Next, we need to calculate the volume of the inner cylinder, which represents the space that is not occupied by the tape. The inner radius (r) is half of the inner diameter, so r = 5 cm / 2 = 2.5 cm.

The volume of the inner cylinder (V_inner) is:
V_inner = π * r² * h
V_inner = π * (2.5 cm)² * 10000 cm
V_inner = π * 6.25 cm² * 10000 cm
V_inner = 62500π cm³

Now, to find the volume of the tape itself, we subtract the volume of the inner cylinder from the volume of the outer cylinder:

V_tile = V_outer - V_inner
V_tile = 250000π cm³ - 62500π cm³
V_tile = 187500π cm³

Finally, to find the thickness of the tape, we divide the volume of the tape by the surface area of the inner cylinder (since the thickness will be uniform and we are considering the volume that the tape occupies over the inner surface area):

Surface area of the inner cylinder (A_inner) is:
A_inner = 2π * r * h
A_inner = 2π * (2.5 cm) * 10000 cm
A_inner = 50000π cm²

The thickness of the tape (t) is:
t = V_tile / A_inner
t = 187500π cm³ / 50000π cm²
t = 3.75 cm

So, the thickness of the tape is 3.75 cm

Since the thickness of the tape is not a practical length (it's too large and not realistic for a thin tape), we need to reconsider our calculation. The thickness should be in the same order of magnitude as the width of the tape, which is 0.2 cm. Let's correct this by using the correct formula for the volume of the cylindrical shell (the tape):

V_tile = A_inner * t

We can now solve for the thickness (t):

t = V_tile / A_inner
t = 187500π cm³ / (π * 50000 cm * 100 cm)
t = 187500 / (50000 * 100)
t = 0.375 cm

Therefore, the thickness of the tape is 0.375 cm.<|endoftext|> [end of text]
```

**Comment by RonanKMcGovern** [comment]
the fp16 looks a bit better, at least it gets to an answer. How does the
pytorch model answer compare?

**Comment by RachidAR** [comment]
I guess I got lucky with a seed. Phi-3-medium (q4 from https://huggingface.co/bartowski/Phi-3-medium-128k-instruct-GGUF, temp 0.1):

_There is a roll of tape. The tape is 100 meters long when unrolled. When rolled up, the outer diameter is 10 cm, and the inner diameter is 5 cm. How thick is the tape?_

Answer:
```
To find out how thick the tape is, we can use the formula for the volume of a cylindrical shell (the shape that represents the unrolled tape). The formula is V = πh(R^2 - r^2), where h is the height or length of the cylinder, R is the outer radius and r is the inner radius.
First, let's convert all measurements to meters for consistency:
- Length (height) of tape, h = 100 m
- Outer diameter = 10 cm = 0.1 m
- Inner diameter = 5 cm = 0.05 m
Now we can find the outer and inner radii by dividing the diameters by 2:
- Outer radius, R = 0.1 m / 2 = 0.05 m
- Inner radius, r = 0.05 m / 2 = 0.025 m
Now we can plug these values into the formula for volume of a cylindrical shell:
V = πh(R^2 - r^2)
We know that V (volume) is equal to the length times the thickness, so let's call the thickness "t" and set up an equation:
100 m * t = π * 100 m * ((0.05 m)^2 - (0.025 m)^2)
Now we can solve for t:
t = (π * ((0.05 m)^2 - (0.025 m)^2)) / 100
t ≈ (3.1416 * (0.0025 m^2 - 0.000625 m^2)) / 100
t ≈ (3.1416 * 0.001875 m^2) / 100
t ≈ 0.0059 m
So the thickness of the tape is approximately 0.0059 meters, or about 5.9 millimeters.
```



> > Did you try running a 16 bit gguf model and seeing how that performs?
> 
> I tried the prompt on the 128k mini instruct model (f16 and Q4_K_M from https://huggingface.co/AlessandroW/Phi-3-mini-128k-instruct-gguf) and both models performed similar to an older ChatGPT version https://neil.fraser.name/news/2023/02/17/. The quantized model replied:
> 
> ```
> <|user|> There is a roll of tape. The tape is 100 meters long when unrolled. When rolled up, the outer diameter is 10 cm, and the inner diameter is 5 cm. How thick is the tape?<|end|><|assistant|> To find the thickness of the tape, we need to calculate the volume of the tape when it is rolled up and then use that to find the thickness.
> 
> First, let's calculate the volume of the cylinder formed by the outer diameter. The formula for the volume of a cylinder is V = πr²h, where r is the radius and h is the height (or length in this case).
> 
> The outer radius (r_outer) is half of the outer diameter, so r_outer = 10 cm / 2 = 5 cm.
> 
> The length of the tape (h) is 100 meters, but we need to convert it to centimeters because the radius is in centimeters. So, h = 100 meters * 100 cm/meter = 10000 cm.
> 
> Now, let's calculate the volume of the outer cylinder (V_outer):
> V_outer = π * (r_outer)² * h
> V_outer = π * (5 cm)² * 10000 cm
> V_outer = π * 25 cm² * 10000 cm
> V_outer = 250000π cm³
> 
> Next, let's calculate the volume of the inner cylinder (V_inner), which represents the empty space inside the tape. The inner radius (r_inner) is half of the inner diameter, so r_inner = 5 cm / 2 = 2.5 cm.
> 
> The volume of the inner cylinder (V_inner) is:
> V_inner = π * (r_inner)² * h
> V_inner = π * (2.5 cm)² * 10000 cm
> V_inner = π * 6.25 cm² * 10000 cm
> V_inner = 62500π cm³
> 
> The volume of the tape itself (V_tape) is the difference between the outer and are not standard mathematical operations, and thus are not applicable in this context.
> 
> Since the question seems to be asking for the volume of the tape itself, we will subtract the inner volume from the outer volume to find the volume of the tape:
> 
> V_tape = V_outer - V_inner
> V_tape = 250000π cm³ - 62500π cm³
> V_tape = 187500π cm³
> 
> To get the numerical value, we use the approximation π ≈ 3.14159:
> 
> V_tape ≈ 187500 * 3.14159 cm³
> V_tape ≈ 588746.25 cm³
> 
> Therefore, the volume of the tape itself is approximately 588746.25 cubic centimeters.<|endoftext|>
> ```
> 
> The f16 model replied
> 
> ```
> <|user|> There is a roll of tape. The tape is 100 meters long when unrolled. When rolled up, the outer diameter is 10 cm, and the inner diameter is 5 cm. How thick is the tape?<|end|><|assistant|> To find the thickness of the tape, we need to calculate the volume of the tape when it is rolled up and then use that to find the thickness.
> 
> First, let's find the volume of the entire length of the tape if it were a cylinder with the outer diameter. The formula for the volume of a cylinder is V = πr²h, where r is the radius and h is the height (or length in this case).
> 
> The outer radius (R) is half of the outer diameter, so R = 10 cm / 2 = 5 cm.
> 
> The length (h) of the tape is 100 meters, which we need to convert to centimeters because the diameter is in centimeters. 100 meters = 100 * 100 cm = 10000 cm.
> 
> Now, let's calculate the volume of the outer cylinder (V_outer):
> V_outer = π * R² * h
> V_outer = π * (5 cm)² * 10000 cm
> V_outer = π * 25 cm² * 10000 cm
> V_outer = 250000π cm³
> 
> Next, we need to calculate the volume of the inner cylinder, which represents the space that is not occupied by the tape. The inner radius (r) is half of the inner diameter, so r = 5 cm / 2 = 2.5 cm.
> 
> The volume of the inner cylinder (V_inner) is:
> V_inner = π * r² * h
> V_inner = π * (2.5 cm)² * 10000 cm
> V_inner = π * 6.25 cm² * 10000 cm
> V_inner = 62500π cm³
> 
> Now, to find the volume of the tape itself, we subtract the volume of the inner cylinder from the volume of the outer cylinder:
> 
> V_tile = V_outer - V_inner
> V_tile = 250000π cm³ - 62500π cm³
> V_tile = 187500π cm³
> 
> Finally, to find the thickness of the tape, we divide the volume of the tape by the surface area of the inner cylinder (since the thickness will be uniform and we are considering the volume that the tape occupies over the inner surface area):
> 
> Surface area of the inner cylinder (A_inner) is:
> A_inner = 2π * r * h
> A_inner = 2π * (2.5 cm) * 10000 cm
> A_inner = 50000π cm²
> 
> The thickness of the tape (t) is:
> t = V_tile / A_inner
> t = 187500π cm³ / 50000π cm²
> t = 3.75 cm
> 
> So, the thickness of the tape is 3.75 cm
> 
> Since the thickness of the tape is not a practical length (it's too large and not realistic for a thin tape), we need to reconsider our calculation. The thickness should be in the same order of magnitude as the width of the tape, which is 0.2 cm. Let's correct this by using the correct formula for the volume of the cylindrical shell (the tape):
> 
> V_tile = A_inner * t
> 
> We can now solve for the thickness (t):
> 
> t = V_tile / A_inner
> t = 187500π cm³ / (π * 50000 cm * 100 cm)
> t = 187500 / (50000 * 100)
> t = 0.375 cm
> 
> Therefore, the thickness of the tape is 0.375 cm.<|endoftext|> [end of text]
> ```
